File: typescript/packages/bench/src/dataset.ts
/**
 * Deterministic in-memory dataset.
 *
 * Fixed rows, no randomness: every lane of this corpus must produce identical
 * numbers so bench expectations can be hard-coded.
 */

import {
    Customer,
    Device,
    Invoice,
    makeTechnician,
    Money,
    Part,
    Priority,
    RepairOrder,
    RepairStatus,
    type Technician,
} from "@atelier/core";

/** Everything a report, metric or service reads from. */
export class Dataset {
    private constructor(
        readonly customers: readonly Customer[],
        readonly devices: readonly Device[],
        readonly orders: readonly RepairOrder[],
        readonly parts: readonly Part[],
        readonly technicians: readonly Technician[],
        readonly invoices: readonly Invoice[],
    ) {}

    /** The frozen seed. Changing these rows changes bench ground truth. */
    static seeded(): Dataset {
        const customers = [
            Customer.seed(1, "Ada Byron", "ada@example.test", "+32 470 00 00 01"),
            Customer.seed(2, "Grace Hopper", "grace@example.test"),
            Customer.seed(3, "Alan Turing", "alan@example.test", "+32 470 00 00 03"),
        ];

        const devices = [
            new Device(1, 1, "Framework", "13", "SER-0001"),
            new Device(2, 2, "Lenovo", "X1", "SER-0002"),
            new Device(3, 3, "Apple", "MBP 14"),
        ];

        const parts = [
            new Part(1, "SCR-13", 'Screen 13"', new Money(19900), 4),
            new Part(2, "BAT-55", "Battery 55Wh", new Money(8900), 1),
            new Part(3, "KBD-EU", "Keyboard EU", new Money(6400), 7),
            new Part(4, "FAN-A1", "Cooling fan", new Money(2200), 2),
        ];

        const technicians = [makeTechnician(1, "Nel"), makeTechnician(2, "Rik"), makeTechnician(3, "Sam")];

        const orders = [
            RepairOrder.seed(1, 1, 1),
            RepairOrder.seed(2, 2, 2),
            RepairOrder.seed(3, 3, 3),
            RepairOrder.seed(4, 1, 1),
        ];

        orders[0]!.laborMinutes = 120;
        orders[0]!.addPart(parts[0]!, 1);
        orders[0]!.transitionTo(RepairStatus.Diagnosing, "seeder");
        orders[0]!.transitionTo(RepairStatus.Repairing, "seeder");
        orders[0]!.transitionTo(RepairStatus.Completed, "seeder");

        orders[1]!.laborMinutes = 45;
        orders[1]!.priority = Priority.Rush;
        orders[1]!.addPart(parts[1]!, 2);
        orders[1]!.transitionTo(RepairStatus.Diagnosing, "seeder");
        orders[1]!.transitionTo(RepairStatus.AwaitingParts, "seeder");

        orders[2]!.laborMinutes = 90;
        orders[2]!.priority = Priority.Warranty;
        orders[2]!.addPart(parts[2]!, 1);
        orders[2]!.addPart(parts[3]!, 1);
        orders[2]!.transitionTo(RepairStatus.Diagnosing, "seeder");
        orders[2]!.transitionTo(RepairStatus.Repairing, "seeder");

        orders[3]!.laborMinutes = 30;

        const invoices = [new Invoice(1, 1, new Money(34900)), new Invoice(2, 2, new Money(23425))];

        return new Dataset(customers, devices, orders, parts, technicians, invoices);
    }

    /** Orders that have reached a billable end state. */
    completedOrders(): readonly RepairOrder[] {
        return this.orders.filter((order) => !order.isOpen());
    }

    /** Orders still occupying bench space. */
    openOrders(): readonly RepairOrder[] {
        return this.orders.filter((order) => order.isOpen());
    }

    ordersFor(customerId: number): readonly RepairOrder[] {
        return this.orders.filter((order) => order.customerId === customerId);
    }

    part(sku: string): Part | undefined {
        return this.parts.find((part) => part.sku === sku);
    }

    lowStockParts(): readonly Part[] {
        return this.parts.filter((part) => part.isLowStock());
    }

    /** Sum of every issued invoice, in cents. */
    revenueCents(): number {
        return this.invoices.reduce((total, invoice) => total + invoice.total.cents, 0);
    }

    /** Sum of every part line across every order, in cents. */
    partsCostCents(): number {
        return this.orders.reduce((total, order) => total + order.partsSubtotal().cents, 0);
    }

    labourMinutes(): number {
        return this.orders.reduce((total, order) => total + order.laborMinutes, 0);
    }
}

File: rust/crates/atelier-bench/src/dataset.rs
//! Deterministic in-memory dataset.
//!
//! Fixed rows, no randomness: the corpus must produce identical numbers on
//! every machine so bench expectations can hard-code them.

use atelier_core::models::invoice::Invoice;
use atelier_core::models::repair_order::RepairOrder;
use atelier_core::money::Money;
use atelier_core::support::priority::Priority;
use atelier_core::support::status::RepairStatus;
use atelier_core::{Customer, Device, Part, Technician};
use serde::Serialize;

/// Everything a report, metric or service reads from.
#[derive(Debug, Clone, Default, Serialize)]
pub struct Dataset {
    pub customers: Vec<Customer>,
    pub devices: Vec<Device>,
    pub orders: Vec<RepairOrder>,
    pub parts: Vec<Part>,
    #[serde(skip)]
    pub technicians: Vec<Technician>,
    pub invoices: Vec<Invoice>,
}

impl Dataset {
    /// The frozen seed. Changing these rows changes bench ground truth.
    pub fn seeded() -> Self {
        let customers = vec![
            Customer::new(1, "Ada Byron", "ada@example.test").with_phone("+32 470 00 00 01"),
            Customer::new(2, "Grace Hopper", "grace@example.test"),
            Customer::new(3, "Alan Turing", "alan@example.test").with_phone("+32 470 00 00 03"),
        ];

        let devices = vec![
            Device::new(1, 1, "Framework", "13").with_serial("SER-0001"),
            Device::new(2, 2, "Lenovo", "X1").with_serial("SER-0002"),
            Device::new(3, 3, "Apple", "MBP 14"),
        ];

        let parts = vec![
            Part::new(1, "SCR-13", "Screen 13\"", Money(19_900)).with_stock(4),
            Part::new(2, "BAT-55", "Battery 55Wh", Money(8_900)).with_stock(1),
            Part::new(3, "KBD-EU", "Keyboard EU", Money(6_400)).with_stock(7),
            Part::new(4, "FAN-A1", "Cooling fan", Money(2_200)).with_stock(2),
        ];

        let technicians = vec![
            Technician::new(1, "Nel"),
            Technician::new(2, "Rik"),
            Technician::new(3, "Sam"),
        ];

        let mut orders = vec![
            RepairOrder::new(1, 1, 1).with_labor(120),
            RepairOrder::new(2, 2, 2)
                .with_priority(Priority::Rush)
                .with_labor(45),
            RepairOrder::new(3, 3, 3)
                .with_priority(Priority::Warranty)
                .with_labor(90),
            RepairOrder::new(4, 1, 1).with_labor(30),
        ];

        orders[0].add_part(&parts[0], 1);
        orders[0].transition_to(RepairStatus::Diagnosing, "seeder");
        orders[0].transition_to(RepairStatus::Repairing, "seeder");
        orders[0].transition_to(RepairStatus::Completed, "seeder");

        orders[1].add_part(&parts[1], 2);
        orders[1].transition_to(RepairStatus::Diagnosing, "seeder");
        orders[1].transition_to(RepairStatus::AwaitingParts, "seeder");

        orders[2].add_part(&parts[2], 1);
        orders[2].add_part(&parts[3], 1);
        orders[2].transition_to(RepairStatus::Diagnosing, "seeder");
        orders[2].transition_to(RepairStatus::Repairing, "seeder");

        let invoices = vec![
            Invoice::new(1, 1, Money(34_900)),
            Invoice::new(2, 2, Money(23_425)),
        ];

        Self {
            customers,
            devices,
            orders,
            parts,
            technicians,
            invoices,
        }
    }

    /// Orders that have reached a billable end state.
    pub fn completed_orders(&self) -> Vec<&RepairOrder> {
        self.orders
            .iter()
            .filter(|order| !order.status.is_open())
            .collect()
    }

    /// Orders still occupying bench space.
    pub fn open_orders(&self) -> Vec<&RepairOrder> {
        self.orders.iter().filter(|order| order.is_open()).collect()
    }

    pub fn orders_for(&self, customer_id: u32) -> Vec<&RepairOrder> {
        self.orders
            .iter()
            .filter(|order| order.customer_id == customer_id)
            .collect()
    }

    pub fn part(&self, sku: &str) -> Option<&Part> {
        self.parts.iter().find(|part| part.sku == sku)
    }

    pub fn low_stock_parts(&self) -> Vec<&Part> {
        self.parts
            .iter()
            .filter(|part| part.is_low_stock())
            .collect()
    }

    /// Sum of every issued invoice, in cents.
    pub fn revenue_cents(&self) -> i64 {
        self.invoices
            .iter()
            .map(|invoice| invoice.total.cents())
            .sum()
    }

    /// Sum of every part line across every order, in cents.
    pub fn parts_cost_cents(&self) -> i64 {
        self.orders
            .iter()
            .map(|order| order.parts_subtotal().cents())
            .sum()
    }

    pub fn labour_minutes(&self) -> u32 {
        self.orders.iter().map(|order| order.labor_minutes).sum()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn seed_is_frozen() {
        let data = Dataset::seeded();

        assert_eq!(data.orders.len(), 4);
        assert_eq!(data.completed_orders().len(), 1);
        assert_eq!(data.revenue_cents(), 58_325);
        assert_eq!(data.low_stock_parts().len(), 2);
    }
}
